### ISSUE TITLE
[Enh]: Provide (basic) testing tables to downstream libraries

### ISSUE THREAD

**Issue by FBruzzesi** [narwhals-dev/narwhals#739]
### We would like to learn about your use case. For example, if this feature is needed  to adopt Narwhals in an open source project, could you please enter the link to it below?

- This is a long shot, and I am not sure which (if any) and how other libraries do it.
- For sure I could see adoption in scikit-lego

### Please describe the purpose of the new feature or describe the problem to solve.

Provide a standard set of values and functionalities to generate tables for each backend we support

### Suggest a solution if possible.

- Have this as a separate library, let's say `narwhals-testing"
- Have this in an internal module, which we can use for our testing purpose as well

### If you have tried alternatives, please describe them below.

Manually create tables in downstream library :)

**Comment by MarcoGorelli** [comment]
Agree! I think @machow also wanted some testing functionality exposed for Shiny

**Comment by FBruzzesi** [comment]
Adding some tracking here for what in my opinion means to close this PR:

- [x] `nw.testing.asserts` module
    - [x] `assert_series_equal`
    - [x] `assert_frame_equal`
- [ ] `nw.testing.constructors` module to provide fixture(s) to use in test suite of downstream libraries
  - [Discord dicussion](https://discord.com/channels/1235257048170762310/1402677544221937705)
  - My naive attempt: [testing-fixtures branch](https://github.com/narwhals-dev/narwhals/compare/main...testing-fixtures)
  - A more sophisticated approach by @EdAbati: [draft-new-constructor](https://github.com/narwhals-dev/narwhals/compare/main...EdAbati:draft-new-constructor)

**Comment by EdAbati** [comment]
>   - A more sophisticated approach by @EdAbati: [draft-new-constructor](https://github.com/narwhals-dev/narwhals/compare/main...EdAbati:draft-new-constructor)

(There is a fine line between 'more sophisticated' and 'overly complicated' 😅 I hope to finish the draft soon)

**Comment by dangotbanned** [comment]
_Originally posted by @elephaint in [#2804](https://github.com/narwhals-dev/narwhals/issues/2804#issue-3210161851)_

> [!NOTE]
> Edited by @dangotbanned 

> ### Is this feature is needed to adopt Narwhals in an open source project?
> 
> At first [`hierarchicalforecast`](https://nixtlaverse.nixtla.io/hierarchicalforecast/index.html), in the future also our other open-source libraries in the [Nixtla](https://github.com/Nixtla) ecosystem
> ### Describe the problem to solve
> 
> [`pandas.testing.assert_frame_equal`](https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html) and [`polars.testing.assert_frame_equal`](https://docs.pola.rs/api/python/stable/reference/api/polars.testing.assert_frame_equal.html) are really convenient when unit testing functionality. 
> 
> Ideally, Narwhals has it too, so that we can abstract away from the dataframe provided and run the unit test on any desired dataframe engine.
> ### Suggest a solution
> 
> Include wrappers for this testing functionality, or include native Narwhals testing functionality, e.g.
> 
> ```py
> nw.testing.asserts.assert_frame_equal
> ```

**Comment by dangotbanned** [comment]
https://github.com/narwhals-dev/narwhals/issues/739#issuecomment-3225435754

> `nw.constructors` module to provide fixture(s) to use in test suite of downstream libraries

@FBruzzesi did you mean `nw.testing.constructors`?
Since the other stuff is in `nw.testing.asserts` 😄

Either way, I'm keen to see  a child issue for that idea and discuss what's going on in yours and @EdAbati's and branches

**Comment by FBruzzesi** [comment]
Hey @EdAbati @dangotbanned  👋🏼

I just got back to this issue as I was taking a look at a downstream library and I was about to re-implement the same constructors I already did too many times.

I am almost starting off from the implementation @EdAbati did.

So far I managed to get far enough to have a pytest plugin which is available once narwhals is installed and provides the same fixtures we have. I will take a stub at refactoring out test suite next

**Comment by dangotbanned** [comment]
@FBruzzesi finally!!!! 😍😍😍

I forgot to mention but I tried out another spin on the constructors stuff fairly recently:


<details><summary>Show <code>TestBackend</code></summary>
<p>


https://github.com/narwhals-dev/narwhals/blob/7b60db32e084fe0ec4b6308ac4544a4883f40fee/tests/plan/utils.py#L292-L647



</p>
</details> 

I've still got lots of tests not using it yet, but maybe there's a thing or two to steal? 😉

**Comment by dangotbanned** [comment]
@FBruzzesi how would you feel about *"adding"* a [`assert_schema_equal`](https://docs.pola.rs/api/python/stable/reference/api/polars.testing.assert_schema_equal.html)?
I started writing a simple version of it (locally) to use in (https://github.com/narwhals-dev/narwhals/compare/oh-nodes...expr-ir/nw-struct).

But when I went to go and steal your `raise_assertion_error` helper, I discovered [we already have one at home](https://github.com/narwhals-dev/narwhals/blob/78df28f250e2c197bf6fd1e14501283e7e6b6afd/src/narwhals/testing/asserts/frame.py#L234-L280) 😉 

I did a quick and dirty version of exposing `_check_schema_equal` in (https://github.com/narwhals-dev/narwhals/compare/main...feat/testing/assert-schema-equal), WDYT?

**Comment by FBruzzesi** [comment]
Hey @dangotbanned I would say it's in scope and as you mentioned, mostly already implemented 🚀 Let me know if you are planning to open a PR upstream

**Comment by dangotbanned** [comment]
Ah nice one thanks @FBruzzesi

> Let me know if you are planning to open a PR upstream

It's all yours if you beat me to it 😉 

---

One thing I noticed (related to #3667), was that this fast path might have the same issue?
https://github.com/narwhals-dev/narwhals/blob/78df28f250e2c197bf6fd1e14501283e7e6b6afd/src/narwhals/testing/asserts/frame.py#L241-L245
